File: 01-Programming-Languages/Kotlin/Algorithms/Sorting/merge_sort.kt
fun mergeSort(arr: IntArray): IntArray {
    if (arr.size <= 1) return arr

    val mid = arr.size / 2
    val left = mergeSort(arr.copyOfRange(0, mid))
    val right = mergeSort(arr.copyOfRange(mid, arr.size))

    return merge(left, right)
}

fun merge(left: IntArray, right: IntArray): IntArray {
    val result = mutableListOf<Int>()
    var i = 0
    var j = 0

    while (i < left.size && j < right.size) {
        if (left[i] <= right[j]) {
            result.add(left[i])
            i++
        } else {
            result.add(right[j])
            j++
        }
    }

    result.addAll(left.drop(i))
    result.addAll(right.drop(j))

    return result.toIntArray()
}

fun main() {
    val arr = intArrayOf(38, 27, 43, 3, 9, 82, 10)
    println("Original: ${arr.toList()}")
    println("Sorted: ${mergeSort(arr).toList()}")
}

File: 01-Programming-Languages/Ruby/Algorithms/Sorting/merge_sort.rb
def merge_sort(arr)
  return arr if arr.length <= 1

  mid = arr.length / 2
  left = merge_sort(arr[0...mid])
  right = merge_sort(arr[mid...])

  merge(left, right)
end

def merge(left, right)
  result = []
  i = j = 0

  while i < left.length && j < right.length
    if left[i] <= right[j]
      result << left[i]
      i += 1
    else
      result << right[j]
      j += 1
    end
  end

  result + left[i..] + right[j..]
end

arr = [38, 27, 43, 3, 9, 82, 10]
puts "Original: #{arr}"
puts "Sorted: #{merge_sort(arr)}"